### DISCUSSION TITLE
Week 10 - Urban Green Space

### DISCUSSION THREAD

**Discussion by eculler** [earthlab-education/Earth-Analytics-2023-01-Intro#556] [Discussion submissions]
Readings:

- [Green space reduces CO2 in cities](https://earth.jpl.nasa.gov/news/23/a-wall-of-green-a-breath-of-fresh-air/) from JPL. [A Wall of Green, a Breath of Fresh Air. JPL Earth Science. Retrieved November 1, 2023, from https://earth.jpl.nasa.gov/news/23/a-wall-of-green-a-breath-of-fresh-air]
- [Why does one city have higher CO2 values than another?](https://ocov2.jpl.nasa.gov/galleries/videos/#images-7) video from JPL/NASA OCO-2 [Yuen, K. Orbiting Carbon Observatory-2 Videos. Retrieved November 1, 2023, from https://ocov2.jpl.nasa.gov/galleries/videos/#images-7]
- [Some open source code for OCO-2](https://github.com/nasa/RtRetrievalFramework/blob/master/python_experimental/finite_diff_check.py)
- [NASA OPERA Python code](https://github.com/nasa/opera-sds-pcm/blob/develop/product2dataset/product2dataset.py)

Questions:
- What do you think of the science communication in the JPL article and video? Who do you think the audience is? Is it effective? What would you do differently?
- Take a look at the OCO-2 code. What would you do to make it cleaner, more expressive, or easier to read?
- Take a look at the OPERA code. Find a `for` loop. What is the code doing? How does the `for` loop help keep the code DRY?

**Comment by Adriannaluger** [comment]
I think the JPL article is written for the general public. It is written as though not much prior knowledge is necessary and the language is broad. She explains green space and its benefits to the audience as to not assume that they already know even though most educated people sort of already have an idea of this. I think she does this very affectively by showing how Flint is the one city where green space has actually increased over time and then continues to provide another way the city wants to imporve green space. She also added pictures throughout the article which is good to visually see as well. However, I think she could have chosen better pictures for citizens of Flint using the green space which I took was the goal of the last image. I would have rathered to see areas with the most green space being used for example versus and very close up image of a group of people who are outside but we cannot actually tell how much green space is around them and what that looks like. To make the code OCO-2 more expressive you could use more meaningful variable names such as changing the names of things to more details words to be able to tell the audience what it is trying to express easier. We may also want to delete repeated code so it is more concise and simple. As well as deleting unnecessary comments like # Plot stuff which is not useful at all. Instead, only have comments that are informative and necessary. In the for loop code " for met_json_file in glob.iglob(os.path.join(dataset_dir, '*.met.json')):
            # Remove the individual .met.json files after they've been merged
            os.unlink(met_json_file)"
It is selecting all the .met.json from the directory and then os.unlink deletes the files that are specified by the given path after they are merged somewhere else in the code. The code continues to be DRY here because it does not repeat anything within the loop, which is good and helps to keep the code concise and maintainable for the future.

**Reply by jsmi374** [reply]
I agree with @adriannaluger on the target audience. Seems like a younger, general public target audience. This is an important issue, and I am happy that they are trying to relate to the general population, but this article seems disjointed and a little empty. As a scientist, I want some numbers and perhaps a map of a neighborhood showing NDVI or vegetated areas before and after green space projects. The author mentions this, but the reader is left to imagine. 

The video is short , to the point, and easy to follow. I like this format, it allows for the audience to quickly tune in and access the information the creator is trying to get across. A length of the video is perfect for getting someone interested. Perhaps, a link to more articles or places to learn more about this topic would be useful for the viewer.

**Reply by peterkobylarz** [reply]
I agree that the first article is written for the general public. It is interesting that an organization like JPL, which is very technically oriented, is publishing articles intended for a non-technical audience, but it makes sense and is encouraging that their science communication program is aimed at a spectrum of audiences, rather than just scientific peers. 

I would have liked the article touch more on the environmental justice portion of the topic of urban greening. You can tell that they are hinting at that topic but fail to actually discuss many of the other environmental and health benefits of green space or the lack thereof, besides air quality. For instance, global heatwaves over the last few years have really highlighted how at risk populations who live in urban areas without trees are much more susceptible to harm during these events than other areas in the same city with more trees.

**Comment by jsmi374** [comment]
OCO-2 code needs to be more expressive. The readme file included in the folder does not provide much information at all. Here is the text included : 
"This is an experimental area that contains code implementing new prototype stuff written in python. This does not get put into the full_physics python library, and we may abandon this stuff at some point." 

New Prototype Stuff??? I understand that this is a working repository, but it would be easier to make the code expressive initially as opposed to returning later to update the code and readme. 

The actual code could be broken up into sections with expressive text to describe "cells" or related parts of code. 


The OPERA code is much cleaner and easier to follow. 

```python
  for output_type in output_types:
        for product in products[output_type].keys():
            logger.info(f"Converting {product} to a dataset")

            dataset_dir = extract.extract(
                os.path.join(product_dir, product),
                settings[extract.PRODUCT_TYPES_KEY],
                os.path.join(product_dir, DATASETS_DIR_NAME),
                extra_met=extra_met,
            )

            hashcheck = products[output_type][product].get("hashcheck", False)

            if hashcheck:
                hash_algo = products[output_type][product].get("hash_algo", DEFAULT_HASH_ALGO)
                create_dataset_checksums(os.path.join(dataset_dir, product), hash_algo)

            created_datasets.add(dataset_dir)
```

This for loop iterates over a list called "output_types". For each "output_type" in this list, it performs several actions on products associated with that output type. This for loop processes a set of products, extracts data, performs a check of the product, and adds the resulting dataset to a directory. The specific actions performed depend on the properties of the products associated with each output type.

Chat GPT helped me describe this!

**Reply by annettebalmes** [reply]
I'm curious to see, this code seemed like an afterthought, with intention to maybe be deleted. Does this mean they ran enough tests, if any at all. I would be hesitant to use this code personally, not only because there are almost no comments or information, but if they don't tell me what it does, how do I know that they even know!

**Comment by peterkobylarz** [comment]
The few comments within the OCO-2 code seem more like the coder's inner thoughts rather than documentation of what is actually happening in the code. I think examples like this are a good reminder to resist the temptation to use comment as our internal dialogue, and make sure that each comment is communicating something useful to whoever is reading them. The massive for loop in the second half of the code has no comment at all explaining what it is doing. At the very least, this large part of the code needs some more explanation.

The OPERA code definitely seems more well documented and expressive. It also seems more DRY and efficient. There are numerous for loops and methods defined in the code while serve to automate repeatable tasks and avoid repetition in the code.

**Comment by ellenalamont17** [comment]
**JPL Article and Video Comments**

From a science communication perspective, the JPL article is excellently crafted. They use visualization and metaphor to help the reader understand and visualize the setting, topic, and difficult concepts while painting a relatable story to appeal to the readers emotions and encourage continued reading. They take an abstract topic like CO2, something considered invisible, and demonstrate it not only as an observable phenomenon but also as an observable problem of concern. And while the article does go a little off topic in places, they always bring it back around to the central point and message - that green spaces help combat carbon emissions and should be more widely adopted in urban landscapes. If I were to do things differently, I might include a description about the role of plants in mitigating CO2 (how do they do it). I would also include some visualizations from the satellite observations to show the difference in CO2 between spring and fall to drive the point home. 

Similarly, I found the video a little lackluster. They do demonstrate differences in CO2 signatures between two cities and explain why they are different, But I would rather have had a follow-up comparison in the same city, either through time or throughout the year to better understand the range of CO2 in a city. They also ended the video before actually describing the various reason a city might high/produce CO2. Can we compare different cities in similar environments/climates but with differing amounts of greenspace? Can we compare neighborhoods in the same city with and without greenspaces? These would have better demonstrated the point they were trying to make and yielded meaningful visualizations.

---

**OCO-2 Code Comments**

I found the OCO-2 code very difficult to follow. There is nearly zero commenting in the code (except technical specifics) making it very difficult to discern what is actually happening in each segment. (They used "# Plot Stuff "...really???). Even the readme file is lacking and over uses the word "stuff". What "stuff" are you referring too??? They did at least format the code into blocks to break up the segments. The variable names, while generally generic, are specific enough to understand what each is supposed to represent. Some of the variable names are far too long (e.g. lua_config.reader.temperature.levels.apriori). Each py file would benefit from the addition of some descriptive text at the beginning of each file as well. 

**Opera Code Comments**

The for loop I've selected comes from the product2dataset folder -> product2dataset.py file. The purpose of this for loop is to create a list of datasets from the various PGE products (directory of files). The loop is essentially extracting the directory of each product file in the product directory, determining a name for the dataset from the product directory name, and turning the files into individual named datasets. A list is generated of the directories of each dataset. This helps to keep the code dry by allowing a single string of code to cycle through numerous lines of text in a directory to perform repetitive actions instead of rewriting the code for each new individual dataset.

>     # Create the datasets
    created_datasets = set()
    output_types = [PRIMARY_KEY, OPTIONAL_KEY]

    for output_type in output_types:
        for product in products[output_type].keys():
            logger.info(f"Converting {product} to a dataset")

            dataset_dir = extract.extract(
                os.path.join(product_dir, product),
                settings[extract.PRODUCT_TYPES_KEY],
                os.path.join(product_dir, DATASETS_DIR_NAME),
                extra_met=extra_met,
            )

            hashcheck = products[output_type][product].get("hashcheck", False)

            if hashcheck:
                hash_algo = products[output_type][product].get("hash_algo", DEFAULT_HASH_ALGO)
                create_dataset_checksums(os.path.join(dataset_dir, product), hash_algo)

            created_datasets.add(dataset_dir)

**Reply by annettebalmes** [reply]
I also took a look at some Opera code and I think you can really see that different people worked on different parts of this data. Your loop doesn't even have a single comment, that's crazy. I looked and saw that everyone did something differently. If not even NASA has a good policy on writing concise, descriptive code, what hope is there for the rest of us working on code as a group. Great job trying to figure out what this code is doing!

**Comment by annettebalmes** [comment]
I took a look at the Opera code, specifically, this function with a for loop:

```python

def merge_dataset_met_json(datasets_parent_dir: str, extra_met: dict) -> Tuple[int, dict]:
    """Merges all the dataset *.met.json metadata into a single dataset metadata dict that can be subsequently saved as *.met.json.
    Returns a tuple of the combined product file sizes and the merged dataset metadata dict.

    :param datasets_parent_dir: the ancestral parent directory of all *.met.json filepaths that should be merged.
    :param extra_met: extra product metadata. This is removed from the dict and added to the dataset metadata.
                      This dict is updated with additional properties from the source dataset *.met.json files.
                      Such properties are prevented from appearing in the merged metadata to prevent duplication.
    """
    dataset_met_json = {"Files": []}
    combined_file_size = 0
    for met_json_file in glob.iglob(os.path.join(datasets_parent_dir, '**/*.met.json'), recursive=True):
        with open(met_json_file, 'r') as infile:
            met_json = json.load(infile)
            combined_file_size += int(met_json["FileSize"])

            # Extract a copy of the "Product*" key/values to include at the top level
            # They should be the same values for each file in the dataset
            product_keys = list(filter(lambda key: key.startswith("Product") or key == "dataset_version", met_json.keys()))

            for product_key in product_keys:
                extra_met[product_key] = met_json[product_key]
                met_json.pop(product_key)

            dataset_met_json["Files"].append(met_json)


    return combined_file_size, dataset_met_json
```

The name gives a pretty good indicator of what it might do and a good description under the function name, but the rest of the code doesn't have many comments to describe what it does, and I think we can do better.

Here's what I believe the loop does:

1.     The loop uses glob.iglob to iterate over all .met.json files in the specified directory (datasets_parent_dir) and its subdirectories (recursive=True).
2.     For each .met.json file, the code opens the file (with open(met_json_file, 'r') as infile) and loads its contents into a dictionary (met_json) using json.load(infile).
3.     The code increments the combined_file_size variable by the value of the "FileSize" key in the loaded metadata (combined_file_size += int(met_json["FileSize"])).
4.     The code identifies keys in the loaded metadata that start with "Product" or are equal to "dataset_version" and extracts their values to include at the top level of the extra_met dictionary. These keys are then removed from the individual met_json dictionaries.
5.     The modified met_json dictionary is appended to the "Files" list in the dataset_met_json dictionary.

I asked ChatGPT to rewrite the function with more descriptive comments, let me know what you think, is it too many? Did Chat get anything wrong? What changes would you make?

```python
def merge_dataset_met_json(datasets_parent_dir: str, extra_met: dict) -> Tuple[int, dict]:
    """Merges all the dataset *.met.json metadata into a single dataset metadata dict that can be subsequently saved as *.met.json.
    Returns a tuple of the combined product file sizes and the merged dataset metadata dict.

    :param datasets_parent_dir: The ancestral parent directory of all *.met.json filepaths that should be merged.
    :param extra_met: Extra product metadata. This is removed from the dict and added to the dataset metadata.
                      This dict is updated with additional properties from the source dataset *.met.json files.
                      Such properties are prevented from appearing in the merged metadata to prevent duplication.
    """
    # Initialize an empty dictionary to store the merged dataset metadata
    dataset_met_json = {"Files": []}

    # Initialize a variable to track the combined file size of all files in the dataset
    combined_file_size = 0

    # Iterate over all .met.json files in the specified directory and its subdirectories
    for met_json_file in glob.iglob(os.path.join(datasets_parent_dir, '**/*.met.json'), recursive=True):
        with open(met_json_file, 'r') as infile:
            # Open the current .met.json file and read its content

            # Load the metadata from the file into a dictionary
            met_json = json.load(infile)

            # Aggregate the file sizes for all files in the dataset
            combined_file_size += int(met_json["FileSize"])

            # Extract a copy of the "Product*" key/values to include at the top level
            # These values should be the same for each file in the dataset
            product_keys = list(filter(lambda key: key.startswith("Product") or key == "dataset_version", met_json.keys()))

            for product_key in product_keys:
                # Extract product-related keys from individual file metadata
                extra_met[product_key] = met_json[product_key]

                # Remove the product-related keys from individual file metadata
                met_json.pop(product_key)

            # Append the modified individual file metadata to the dataset metadata
            dataset_met_json["Files"].append(met_json)

    # Return a tuple of the combined file sizes and the merged dataset metadata
    return combined_file_size, dataset_met_json
```
